[com_fields] Reload task belongs to the component controller

laoneo opened the pull request:
Pull Request for Issue #15995.

### Summary of Changes
When a category changes for an article the page got reloaded to determine the custom fields. Actually this action got delegated to the com_fields controller. On that point the data is unvalidated stored in the session. This can lead to time shifts when the Joomla timezone is not UTC, more information can be found in issue #15995.

This pr reloads the page trough the component controller and not the com_fields one.

### Testing Instructions
- Set the global Joomla timezone to a different one than UTC, for example Berlin
- Create two article categories
- Create a calendar custom field
- Set show time to yes in the field
- Save the field
- Create an article
- Change the category of the article

### Expected result
The time in the custom field remains.

### Actual result
The time in the custom field shifts the same hours as the Joomla timezone has.

laoneo committed:
Reload task belongs to the component controller

franz-wohlkoenig commented:
I have tested this item :red_circle: unsuccessfully on 64d9bb2ff51aba0b6d7bd40a688de3301a5e865b<br /><br />Set Time-Zone to "Berlin", 
applied PR, 
change Category > +2h in Calendar-Field
changed back to 1st Category > additional 2h (4 Hours)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16665">issues.joomla.org/tracker/joomla-cms/16665</a>.</sub>

laoneo commented:
Do the publishing dates also shift?

franz-wohlkoenig commented:
I have tested this item :white_check_mark: successfully on 64d9bb2ff51aba0b6d7bd40a688de3301a5e865b<br /><br />Test again, all works fine now.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16665">issues.joomla.org/tracker/joomla-cms/16665</a>.</sub>

wilsonge commented:
We need to be careful of ACL here - this is going to expose public data if accessed directly from a URL

laoneo commented:
Really, it uses the same code and checks as the save function. Which line do you think can be dangerous?

Bakual commented:
> Really, it uses the same code and checks as the save function. Which line do you think can be dangerous?

Maybe I miss something but I don't see any ACL checks at all here.
The save method uses `$this->allowSave($data, $key)` to check if saving is allowed which I don't see in your reload method.
We may also have to use `checkEditId($context, $id)` to verify that the item was correctly opened before.

laoneo commented:
Why is the allowSave check needed, it doesn't save, it just reloads the page to an url like /cms/administrator/index.php?option=com_content&view=article&layout=edit&id=1. The checkedit id is done in the controller itself. I mean at the end it performs the same action as when you want to edit the article. If you paste that url directly into the browser and you didn't go trough the edit task, then it kicks you back to the list with a warning to not get directly to the edit page. I'v just tested it and it worked as expected.
If I'm now not completely wrong then I don't understand your points here. What else needs to be checked?

Bakual commented:
> Why is the allowSave check needed, it doesn't save, it just reloads the page to an url like /cms/administrator/index.php?option=com_content&view=article&layout=edit&id=1.

It does save the data into the session. It at least feels wrong to have a controller task without any ACL check. Which means anyone could run it. It may not be an issue by itself, but it may bite us back and there is no reason this task should be available unauthorised. Using allowSave or allowEdit (if that fits better due to not saving into db) is a save check we can and should do imho. I don't see a reason to not do it.

As for checkEditId, you're right. I just remembered we check it but didn't remember where exactly :smile:

wilsonge commented:
Does this not allow accessing data from any JControllerForm instance (including in the frontend (e.g. com_contact)) without checking the published/acl levels for the item though?

laoneo commented:
I don't se how, it only stores the jform data into the session.

laoneo committed:
Add an extra check if the record can be edited.

mbabker commented:
It is still a POST request receiving user data.  It should therefore be properly validated and access controlled.  Especially if that data stored into the session can be used in a follow on request (which HAS been used in security issues we have fixed in the last 2 years).

wilsonge commented:
The session is data that the user controls. That means the user has the data which they shouldn't have been allowed to access (for ACL reasons) - and even if it isn't displayed on the page. It's trivial to get the session data if you wanted it.

laoneo commented:
Does that help https://github.com/joomla/joomla-cms/pull/16665/commits/d10e8fc78bada54bdd6b72feffbb0c6b4c4420d5?

Bakual commented:
Imho, the ACL check is now fine.

Looking at the redirects, why don't you just use `$this->redirect()`? Also `$app->close()` is useless even when you use `$app->redirect()` directly, since that method already closes the app. Or do I miss something?

laoneo commented:
Just to be sure. Probably not needed.

Bakual commented:
If there is no specific reason to use `$app->redirect()` directly, can you just use `$this->setRedirect($url, $msg, $type)` and `$this->redirect()` then? So we are consistent with how we process redirects in other places in the controllers.

laoneo committed:
Use the controller redirect functions

laoneo commented:
True, done.

wilsonge commented:
RE ACL: I'm honestly unsure given what this function does whether it should be edit or read acl permissions we should be using. But this looks significantly better now :)

laoneo commented:
It should check edit, as it belongs to the edit action.

laoneo committed:
Merge remote-tracking branch 'remotes/upstream/staging' into cf/reload-form

# Conflicts:
#	administrator/components/com_fields/helpers/fields.php

infograf768 commented:
@laoneo 
I can't get the calendar field anymore
( ! ) Warning: Invalid argument supplied for foreach() in /Applications/MAMP/htdocs/testnew/trunkgitnew/libraries/cms/helper/content.php on line 74

infograf768 commented:
![fields](https://user-images.githubusercontent.com/869724/27950516-ed212420-6301-11e7-8ebf-f6448098c1f1.gif)

infograf768 commented:
Forgot to delete `administrator/components/com_fields/controllers/field.php`

After deleting it, I get:
`An error has occurred.`
` 0 Invalid controller: name='field', format='' `

infograf768 commented on administrator/components/com_admin/script.php:
```diff
@@ -1723,6 +1723,8 @@ public function deleteUnexistingFiles()
 			'/administrator/components/com_messages/layouts/toolbar',
 			'/administrator/components/com_messages/layouts',
+			// Joomla! __DEPLOY_VERSION__
+			'/components/com_fields/controllers',
```
this targetting a folder. Should be the file.
`'/components/com_fields/controllers/field.php',`

laoneo replied:
This folder has only one file, so better to remove it completely.

infograf768 replied:
better do both. see https://github.com/Digital-Peak/joomla-cms/blob/feb148b97dbab493f21a02a289ddf8d06df03692/administrator/components/com_admin/script.php#L1599

infograf768 replied:
https://github.com/Digital-Peak/joomla-cms/blob/feb148b97dbab493f21a02a289ddf8d06df03692/administrator/components/com_admin/script.php#L1599
better do both

laoneo replied:
Why?

infograf768 replied:
As comment says, deleting only the folder may not work when ftp is used.
It is safer to at least delete the file.

laoneo committed:
Move type field to correct reload function

laoneo committed:
Only use filter function as we are not validating the data on reload

If a new article is created and the title is not set and the category has changed, then the validated data is false, will not be set and the admin is loosing all its entered data. So we need to filter here only.

laoneo commented:
Good catch. Fixed it in the last commits.

infograf768 commented:
Works now. Remains the necessary changes in script.php

laoneo commented:
Ok, done.

laoneo committed:
Add the file to the remove script

infograf768 commented:
I have tested this item :white_check_mark: successfully on fb925e7cd3b0304cb778297cf2d08918670c9994<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16665">issues.joomla.org/tracker/joomla-cms/16665</a>.</sub>

franz-wohlkoenig commented:
I have tested this item :white_check_mark: successfully on fb925e7cd3b0304cb778297cf2d08918670c9994<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16665">issues.joomla.org/tracker/joomla-cms/16665</a>.</sub>

franz-wohlkoenig commented:
RTC after two successful tests.

laoneo commented:
Thanks guys!
